File: 2020/Go/day-06.go
package main
import (
    "bufio"
    "fmt"
    "os"
    "strings"
)

// Part 1
func checkMembership(input, element string) bool {
    for _, i := range input {
        value := string(i)
        if (value == element) { return true }
    }
    return false
}

func part1(groups [][]string) (count int) {
    for _, group := range groups {
        var answers string
        for _, individualResponse := range group {
            for _, question := range individualResponse {
                if !checkMembership(answers, string(question)) {
                    answers += string(question)
                }
            }
        }
        count += len(answers)
    }
    return
}

// Part 2
func part2(groups [][]string) (count int) {
    for _, group := range groups {
        answers := strings.Join(group, "")
        var answersChecked string
        for _, i := range answers {
            question := string(i)
            if (strings.Count(answers, question) == len(group) && !checkMembership(answersChecked, question)) {
                answersChecked += question
                count += 1
            }
        }
    }
    return
}

func main() {
    // Read input
    file, _ := os.Open("input.txt")
    defer file.Close()

    scanner := bufio.NewScanner(file)
    var input []string

    for scanner.Scan() {
        input = append(input, scanner.Text())
    }
    // Add EOF newline
    input = append(input, "")

    var groups [][]string
    var tempGroup []string
    for _, i := range input {
        if (i == "") {
            groups = append(groups, tempGroup)
            tempGroup = nil
        } else {
            tempGroup = append(tempGroup, i)
        }
    }

    // Print answers
    fmt.Println(part1(groups))
    fmt.Println(part2(groups))
}

File: 2020/python/day-06.py
with open("input.txt") as input_file:
    _input = input_file.read().splitlines()
    # Add EOF newline
    _input.append("")

    # Format input
    groups = []
    temp_group = []

    for i in _input:
        if (i == ""):
            groups.append(temp_group)
            temp_group = []
        else:
            temp_group.append(i)

# Part 1
def part_1(groups):
    count = 0
    for group in groups:
        answers = ""
        for individual_response in group:
            for question in individual_response:
                if (question not in answers):
                    answers += question
        count += len(answers)
    return count

def part_2(groups):
    count = 0
    for group in groups:
        answers = "".join(group)
        answers_checked = ""
        for question in answers:
            if (answers.count(question) == len(group) and question not in answers_checked):
                answers_checked += question
                count += 1
    return count

# Print answers
print(part_1(groups))
print(part_2(groups))
# [Finished in 301ms]